// telefunc/dist/cjs/node/server/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.__decorateTelefunction = exports.onBug = exports.shield = exports.Abort = exports.provideTelefuncContext = exports.getContext = exports.telefuncConfig = exports.config = exports.telefunc = void 0;
var telefunc_1 = require("./telefunc");
Object.defineProperty(exports, "telefunc", { enumerable: true, get: function () { return telefunc_1.telefunc; } });
const serverConfig_1 = require("./serverConfig");
Object.defineProperty(exports, "config", { enumerable: true, get: function () { return serverConfig_1.config; } });
Object.defineProperty(exports, "telefuncConfig", { enumerable: true, get: function () { return serverConfig_1.config; } });
var getContext_1 = require("./getContext");
Object.defineProperty(exports, "getContext", { enumerable: true, get: function () { return getContext_1.getContext; } });
Object.defineProperty(exports, "provideTelefuncContext", { enumerable: true, get: function () { return getContext_1.provideTelefuncContext; } });
var Abort_1 = require("./Abort");
Object.defineProperty(exports, "Abort", { enumerable: true, get: function () { return Abort_1.Abort; } });
var shield_1 = require("./shield");
Object.defineProperty(exports, "shield", { enumerable: true, get: function () { return shield_1.shield; } });
var onBug_1 = require("./runTelefunc/onBug");
Object.defineProperty(exports, "onBug", { enumerable: true, get: function () { return onBug_1.onBug; } });
var decorateTelefunction_1 = require("./runTelefunc/decorateTelefunction");
Object.defineProperty(exports, "__decorateTelefunction", { enumerable: true, get: function () { return decorateTelefunction_1.decorateTelefunction; } });
const utils_1 = require("../utils");
assertServerSide();
function assertServerSide() {
    var _a;
    const isBrowser = typeof window !== 'undefined' && 'innerHTML' in (((_a = window === null || window === void 0 ? void 0 : window.document) === null || _a === void 0 ? void 0 : _a.body) || {});
    (0, utils_1.assertUsage)(!isBrowser, [
        'You are loading the `telefunc` module in the browser, but',
        'the `telefunc` module can only be imported in Node.js.',
    ].join(' '));
}
